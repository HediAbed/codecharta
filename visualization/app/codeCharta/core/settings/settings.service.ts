import { Settings, Vector3d, RenderMode, MapColors, RecursivePartial } from "../../codeCharta.model"
import _ from "lodash"
import { IRootScopeService, IAngularEvent } from "angular";

export interface SettingsServiceSubscriber {
	onSettingsChanged(settings: Settings, event: IAngularEvent)
}

export class SettingsService {

    public static SELECTOR = "settingsService"
    private static SETTINGS_CHANGED_EVENT = "settings-changed";
	private static MIN_MARGIN = 15

	private settings: Settings 
	throttledBroadcast: ()=>void;

	constructor(private $rootScope) {
		this.settings = this.getDefaultSettings()
		this.throttledBroadcast = _.throttle(() => this.$rootScope.$broadcast(SettingsService.SETTINGS_CHANGED_EVENT, this.settings), 400)
	}

	public getSettings(): Settings {
		return this.settings
	}

	public updateSettings(update: RecursivePartial<Settings>) {
		// TODO wo und wann this.settings.margin = this.computeMargin(settings); ?
		this.settings = _.merge(this.settings, update)
		this.throttledBroadcast()
	}

	/* TODO this needs to be called at other places, this cannot be here
	private potentiallyUpdateColorRange(update: RecursivePartial<Settings>): RecursivePartial<Settings> {
		if (update.mapSettings.colorMetric) {
			update.appSettings.neutralColorRange = this.getAdaptedRange(update.mapSettings.colorMetric)
		}
		return update
	}*/

	public getDefaultSettings(): Settings {
		const mapColors: MapColors = {
			positive: "#69AE40",
			neutral: "#ddcc00",
			negative: "#820E0E",
			selected: "#EB8319",
			defaultC: "#89ACB4",
			positiveDelta: "#69FF40",
			negativeDelta: "#ff0E0E",
			base: "#666666",
			flat: "#AAAAAA",
			lightGrey: "#DDDDDD",
			angularGreen: "#00BFA5",
			markingColors: ["#FF1D8E", "#1d8eff", "#1DFFFF", "#8eff1d", "#8e1dff", "#FFFF1D"]
		}

		const scaling: Vector3d = {
			x: 1,
			y: 1,
			z: 1
		}

		const camera: Vector3d = {
			x: 0,
			y: 300,
			z: 1000
		}

		let settings: Settings = {
			mapSettings: {
				renderMode: RenderMode.Single,
				areaMetric: "",
				heightMetric: "",
				colorMetric: "",
				focusedNodePath: null,
				searchedNodePaths: [],
				searchPattern: null,
				attributeTypes: {},
				blacklist: [],
				edges: []
			},
			appSettings: {
				amountOfTopLabels: 1,
				scaling: scaling,
				camera: camera,
				margin: SettingsService.MIN_MARGIN,
				neutralColorRange: null,
				deltaColorFlipped: false,
				enableEdgeArrows: true,
				hideFlatBuildings: true,
				maximizeDetailPanel: false,
				invertHeight: false,
				dynamicMargin: true,
				isWhiteBackground: false,
				whiteColorBuildings: false,
				mapColors: mapColors
			}
		}

		return settings
	}

	public static subscribe($rootScope: IRootScopeService, subscriber: SettingsServiceSubscriber) {
		$rootScope.$on(SettingsService.SETTINGS_CHANGED_EVENT, (event, data) => {
			subscriber.onSettingsChanged(data, event)
		})
	}

	/* TODO someone else does this
    public onDataChanged(data: DataModel) {

        if(data.metrics && data.renderMap && data.revisions) {
            this.settings.map = data.renderMap;
            this.settings.blacklist = data.renderMap.blacklist;

            if (data.metrics.indexOf(this.settings.areaMetric) === -1) {
                this.settings.areaMetric = this.getMetricByIdOrLast(0, data.metrics);
            }

            if (data.metrics.indexOf(this.settings.heightMetric) === -1) {
                this.settings.heightMetric = this.getMetricByIdOrLast(1, data.metrics);
            }

            if (data.metrics.indexOf(this.settings.colorMetric) === -1) {
                this.settings.colorMetric = this.getMetricByIdOrLast(2, data.metrics);
            }

            this.onSettingsChanged();
        }

    }*/

	/* TODO someone else does this, updateSettingsMethod should offer silent flag (no event triggered)
    public onCameraChanged(camera: PerspectiveCamera) {
        if (
            this.settings.camera.x !== camera.position.x ||
            this.settings.camera.y !== camera.position.y ||
            this.settings.camera.z !== camera.position.z
        ) {
            this.settings.camera.x = camera.position.x;
            this.settings.camera.y = camera.position.y;
            this.settings.camera.z = camera.position.z;
            // There is no component in CC which needs live updates when camera changes. Broadcasting an
            // onSettingsChanged Event would cause big performance issues
            // this.onSettingsChanged();
        }
    }*/

	/* TODO move 
    public updateSettingsFromUrl() {

        let iterateProperties = (obj, prefix) => {
            for (let i in obj) {
                if (obj.hasOwnProperty(i) && i !== "map" && i) {

                    if (typeof obj[i] === "string" || obj[i] instanceof String) {
                        //do not iterate over strings
                    } else {
                        iterateProperties(obj[i], i + ".");
                    }

                    const res = this.urlService.getParam(prefix + i);

                    if (res) {

                        if (res == "true") {
                            obj[i] = true;
                        } else if (res == "false") {
                            obj[i] = false;
                        } else if (res === 0 || res) {

                            let val = parseFloat(res);

                            if (isNaN(val)) {
                                val = res;
                            }

                            obj[i] = val;
                        } else {
                            obj[i] = res;
                        }

                        // if we work with strings here it can cause errors in other parts of the app, check with console.log(typeof obj[i], obj[i]);
                    }

                }
            }
        };

        iterateProperties(this.settings, "");


        this.onSettingsChanged();

    }
    */

    /* TODO move
    	//private static MAX_MARGIN = 100

    public computeMargin(s: Settings = this.settings): number {
        if (s.map !== null && s.dynamicMargin) {
            let leaves = hierarchy<CodeMapNode>(s.map.nodes).leaves();
            let numberOfBuildings = 0;
            let totalArea = 0;

            leaves.forEach((node: HierarchyNode<CodeMapNode>) => {
                numberOfBuildings++;
                if(node.data.attributes && node.data.attributes[s.areaMetric]){
                    totalArea += node.data.attributes[s.areaMetric];
                }
            });

            let margin: number = SettingsService.MARGIN_FACTOR * Math.round(Math.sqrt((totalArea / numberOfBuildings)));
            return Math.min(SettingsService.MAX_MARGIN, Math.max(SettingsService.MIN_MARGIN, margin));
        } else {
            return s.margin;
        }
    }*/
}
