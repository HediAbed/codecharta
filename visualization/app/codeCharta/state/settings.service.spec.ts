import {SettingsService} from "./settings.service";
import {IRootScopeService} from "angular";
import {getService, instantiateModule, NGMock} from "../../../mocks/ng.mockhelper";

describe("app.codeCharta.core.settings", () => {

    let validCodeMap: CodeMap
    let settingsService: SettingsService
    let $rootScope: IRootScopeService

    beforeEach(() => {
        restartSystem();
        rebuildService();
    });

    function restartSystem() {
        instantiateModule("app.codeCharta.state.settings");

        $rootScope = getService<IRootScopeService>("$rootScope")
    }

    function rebuildService() {
        settingsService = new SettingsService($rootScope)
    }

    describe("Get and set url parameter", () => {

        it("should retrieve the correct query param strings", () => {
            settingsService.settings.areaMetric = "areaStuff";
            settingsService.settings.camera.x = 2;
            expect(settingsService.getQueryParamString()).toContain("areaMetric=areaStuff");
            expect(settingsService.getQueryParamString()).toContain("camera.x=2");
            expect(settingsService.getQueryParamString()).toContain("neutralColorRange.from=0");
        });

        it("should update settings from url", NGMock.mock.inject(($location) => {
            $location.url("http://something.de?scaling.x=42&areaMetric=myMetric&scaling.y=0.32");
            settingsService.updateSettingsFromUrl();
            expect(settingsService.settings.scaling.x).toBe(42);
            expect(settingsService.settings.scaling.y).toBe(0.32);
            expect(settingsService.settings.areaMetric).toBe("myMetric");
        }));

        it("should not update settings.map from url", NGMock.mock.inject(($location) => {
            $location.url("http://something.de?map=aHugeMap");
            settingsService.settings.map = "correctMap" as any as CodeMap;
            settingsService.updateSettingsFromUrl();
            expect(settingsService.settings.map).toBe("correctMap");
        }));
    });

    describe("onSettingsChanged", () => {

        it("should react to data-changed events", () => {
            settingsService.onSettingsChanged = jest.fn();

            //enough metrics
            services.$rootScope.$broadcast("data-changed", {
                renderMap: validCodeMap,
                metrics: ["a", "b", "c"],
                revisions: [validCodeMap, validCodeMap]
            });

            expect(settingsService.settings.map.fileName).toBe("file");
            expect(settingsService.settings.areaMetric).toBe("a");
            expect(settingsService.settings.heightMetric).toBe("b");
            expect(settingsService.settings.colorMetric).toBe("c");

            //not enough metrics
            validCodeMap.fileName = "file2";
            services.$rootScope.$broadcast("data-changed", {
                renderMap: validCodeMap,
                metrics: ["a"],
                revisions: [validCodeMap, validCodeMap]
            });

            expect(settingsService.settings.map.fileName).toBe("file2");
            expect(settingsService.settings.areaMetric).toBe("a");
            expect(settingsService.settings.heightMetric).toBe("a");
            expect(settingsService.settings.colorMetric).toBe("a");

            expect(settingsService.onSettingsChanged).toHaveBeenCalledTimes(2);
        });

        it("should react to camera-changed events", () => {
            settingsService.onCameraChanged = sinon.spy();
            services.$rootScope.$broadcast("camera-changed", {});
            expect(settingsService.onCameraChanged.calledOnce);
        });

        it("onCameraChanged should update settings object but not call onSettingsChanged to ensure performance", () => {
            settingsService.onSettingsChanged = sinon.spy();
            settingsService.onCameraChanged({position: {x: 0, y: 0, z: 42}});
            expect(!settingsService.onSettingsChanged.called);
            expect(settingsService.settings.camera.z).toBe(42);
        });

        it("should react to data-changed events and set metrics correctly", () => {
            settingsService.onSettingsChanged = sinon.spy();

            services.$rootScope.$broadcast("data-changed", {
                renderMap: validCodeMap,
                metrics: ["a", "b"],
                revisions: [validCodeMap, validCodeMap]
            });

            expect(settingsService.settings.map.fileName).toBe("file");
            expect(settingsService.settings.areaMetric).toBe("a");
            expect(settingsService.settings.heightMetric).toBe("b");
            expect(settingsService.settings.colorMetric).toBe("b");
            expect(settingsService.onSettingsChanged.calledOnce);
        });
    });
});