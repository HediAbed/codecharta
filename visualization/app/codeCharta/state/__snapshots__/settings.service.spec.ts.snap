// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`settingService getDefaultSettings should match snapshot 1`] = `
Object {
  "appSettings": Object {
    "amountOfEdgePreviews": 1,
    "amountOfTopLabels": 1,
    "camera": Vector3 {
      "x": 0,
      "y": 300,
      "z": 1000,
    },
    "dynamicMargin": true,
    "enableEdgeArrows": true,
    "hideFlatBuildings": true,
    "invertColorRange": false,
    "invertDeltaColors": false,
    "invertHeight": false,
    "isPresentationMode": false,
    "isWhiteBackground": false,
    "mapColors": Object {
      "angularGreen": "#00BFA5",
      "base": "#666666",
      "defaultC": "#89ACB4",
      "flat": "#AAAAAA",
      "lightGrey": "#DDDDDD",
      "markingColors": Array [
        "#FF1D8E",
        "#1d8eff",
        "#1DFFFF",
        "#8eff1d",
        "#8e1dff",
        "#FFFF1D",
      ],
      "negative": "#820E0E",
      "negativeDelta": "#ff0E0E",
      "neutral": "#ddcc00",
      "positive": "#69AE40",
      "positiveDelta": "#69FF40",
      "selected": "#EB8319",
    },
    "maximizeDetailPanel": false,
    "scaling": Vector3 {
      "x": 1,
      "y": 1,
      "z": 1,
    },
    "whiteColorBuildings": false,
  },
  "dynamicSettings": Object {
    "areaMetric": null,
    "colorMetric": null,
    "colorRange": Object {
      "from": null,
      "to": null,
    },
    "distributionMetric": null,
    "edgeMetric": "",
    "focusedNodePath": "",
    "heightMetric": null,
    "margin": null,
    "searchPattern": "",
    "searchedNodePaths": Array [],
  },
  "fileSettings": Object {
    "attributeTypes": Object {
      "edges": Array [],
      "nodes": Array [],
    },
    "blacklist": Array [],
    "edges": Array [],
    "markedPackages": Array [],
  },
  "treeMapSettings": Object {
    "mapSize": 500,
  },
}
`;
